x/stakeibc/keeper/msg_server_instant_redeem_stake_test.go

```diff
@@ -0,0 +1,192 @@
+	"fmt"
+	epochtypes "github.com/Stride-Labs/stride/v5/x/epochs/types"
+	recordtypes "github.com/Stride-Labs/stride/v5/x/records/types"
+	stakeibctypes "github.com/Stride-Labs/stride/v5/x/stakeibc/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	_ "github.com/stretchr/testify/suite"
+)
+
+type InstantRedeemStakeState struct {
+	epochNumber         uint64
+	depositRecordAmount sdkmath.Int
+	hostZone            stakeibctypes.HostZone
+}
+
+type InstantRedeemStakeTestCase struct {
+	user         Account
+	zoneAccount  Account
+	initialState InstantRedeemStakeState
+	validMsg     stakeibctypes.MsgInstantRedeemStake
+}
+
+func (s *KeeperTestSuite) SetupInstantRedeemStake() InstantRedeemStakeTestCase {
+	unbondAmount := sdkmath.NewInt(1_000_000)
+	depositAmount := sdkmath.NewInt(500_000)
+	user := Account{
+		acc:           s.TestAccs[0],
+		atomBalance:   sdk.NewInt64Coin(IbcAtom, 10_000_000),
+		stAtomBalance: sdk.NewInt64Coin(StAtom, 10_000_000),
+	}
+	s.FundAccount(user.acc, user.atomBalance)
+	s.FundAccount(user.acc, user.stAtomBalance)
+
+	zoneAddress := stakeibctypes.NewZoneAddress(HostChainId)
+
+	zoneAccount := Account{
+		acc:           zoneAddress,
+		atomBalance:   sdk.NewInt64Coin(IbcAtom, 10_000_000),
+		stAtomBalance: sdk.NewInt64Coin(StAtom, 10_000_000),
+	}
+	s.FundAccount(zoneAccount.acc, zoneAccount.atomBalance)
+	s.FundAccount(zoneAccount.acc, zoneAccount.stAtomBalance)
+
+	hostZone := stakeibctypes.HostZone{
+		ChainId:        HostChainId,
+		HostDenom:      Atom,
+		IbcDenom:       IbcAtom,
+		RedemptionRate: sdk.NewDec(1.0),
+		Address:        zoneAddress.String(),
+		StakedBal:      unbondAmount,
+	}
+
+	epochTracker := stakeibctypes.EpochTracker{
+		EpochIdentifier: epochtypes.STRIDE_EPOCH,
+		EpochNumber:     1,
+	}
+
+	oldDepositRecord := recordtypes.DepositRecord{
+		Id:                 0,
+		DepositEpochNumber: 0,
+		HostZoneId:         "GAIA",
+		Amount:             depositAmount,
+	}
+
+	initialDepositRecord := recordtypes.DepositRecord{
+		Id:                 1,
+		DepositEpochNumber: 1,
+		HostZoneId:         "GAIA",
+		Amount:             depositAmount,
+	}
+
+	s.App.StakeibcKeeper.SetHostZone(s.Ctx, hostZone)
+	s.App.StakeibcKeeper.SetEpochTracker(s.Ctx, epochTracker)
+	s.App.RecordsKeeper.SetDepositRecord(s.Ctx, oldDepositRecord)
+	s.App.RecordsKeeper.SetDepositRecord(s.Ctx, initialDepositRecord)
+
+	return InstantRedeemStakeTestCase{
+		user:        user,
+		zoneAccount: zoneAccount,
+		initialState: InstantRedeemStakeState{
+			epochNumber:         epochTracker.EpochNumber,
+			depositRecordAmount: depositAmount,
+			hostZone:            hostZone,
+		},
+		validMsg: stakeibctypes.MsgInstantRedeemStake{
+			Creator:  user.acc.String(),
+			HostZone: HostChainId,
+			Amount:   unbondAmount,
+		},
+	}
+}
+
+// TODO: Need to add tests for at least multiple deposit records, non 1.0 Redemption Rates, and probably some other basic scenarios.
+func (s *KeeperTestSuite) TestInstantRedeemStake_Successful() {
```

Comment:
Could we add tests for
- no new funds were staked in this epoch (q: in this case, will the DR exist with value 0 or will it not exist?)
- instant redeem staking the entire amount of the DR, then trying to instant redeem stake again after that (second should fail I think, but I believe the DR should remain, with value 0)

Reply:
Each epoch creates a record, with value of 0.  So if there are no new funds, it is there with a value of 0.  I'll add both cases when I do more scenarios.

Reply:
+1 on testing the DR goes unchanged. Would be good to add a case that redeems 750_000 - it should deplete one DR completely and then deplete the second by 50%.
Also agree with your comment - I'd definitely add a test with a non-1.0 redemption rate!

Reply:
Leaving this open until we add these tests

Reply:
Most of these tests have been added, just missing one case

Reply:
All cases are in there now.